cmd/summary.go

```diff
@@ -35,51 +35,56 @@ var (
 	extFile string
 )
 
-const summaryLongDescription string = `summarize model(s), for example: 
+const summaryLongDescription string = `summarize model(s), for example:
 bbi nonmem summary run001/run001
 bbi nonmem summary run001/run001.lst
 bbi nonmem summary run001/run001.res
+bbi nonmem summary run001/run001 run002/run002
  `
 
 type jsonResults struct {
 	Results []parser.SummaryOutput
-	Errors  []error
+	Errors  []int
+}
+
+func summarizeModel(model string) (parser.SummaryOutput, error) {
+	return parser.GetModelOutput(model, parser.NewModelOutputFile(extFile, noExt),
+		!noGrd, !noShk)
 }
 
 func summary(_ *cobra.Command, args []string) {
```

Comment:
Random input, it might make sense to use Cobras `RunE` vs `Run` directives since they return an error up the stack. That way your `summary` function can just return errors and allow the executing layers calling `Summary.Execute()` to exit at that point. 

Would give you a single control point for the execution versus having to manage individual os exit calls farther down the stack

Reply:
@shairozan Thanks for the input.

So I guess at the minimum that would look something like this:

<details>
<summary>diff</summary>

```
diff --git a/cmd/summary.go b/cmd/summary.go
index cb0a45e..157496e 100644
--- a/cmd/summary.go
+++ b/cmd/summary.go
@@ -16,7 +16,6 @@ package cmd

 import (
    "fmt"
-	"os"
    "runtime"

    parser "github.com/metrumresearchgroup/bbi/parsers/nmparser"
@@ -52,7 +51,7 @@ func summarizeModel(model string) (parser.SummaryOutput, error) {
        !noGrd, !noShk)
 }

-func summary(_ *cobra.Command, args []string) {
+func summary(_ *cobra.Command, args []string) error {
    if debug {
        viper.Debug()
    }
@@ -64,22 +63,22 @@ func summary(_ *cobra.Command, args []string) {
            if Json {
                results = parser.SummaryOutput{ErrorMsg: err.Error()}
            } else {
-				log.Fatal(err)
+				return err
            }
        }
        if Json {
            err = utils.PrintJSON(results)
            if err != nil {
-				log.Fatal(err)
+				return err
            }
            if !results.Success {
-				os.Exit(1)
+				return fmt.Errorf("summarizing model failed: %v", results.ErrorMsg)
            }
        } else {
            results.Summary()
        }

-		return
+		return nil
    }

    type modelResult struct {
@@ -139,7 +138,7 @@ func summary(_ *cobra.Command, args []string) {
    if Json {
        err := utils.PrintJSON(modelResults)
        if err != nil {
-			log.Fatal(err)
+			return err
        }
    } else {
        // not json lets print all successful models first then any errors
@@ -159,8 +158,10 @@ func summary(_ *cobra.Command, args []string) {
        }
    }
    if nerrors > 0 {
-		os.Exit(1)
+		return fmt.Errorf("summarizing %d model(s) failed", nerrors)
    }
+
+	return nil
 }

 func NewSummaryCmd() *cobra.Command {
@@ -168,7 +169,7 @@ func NewSummaryCmd() *cobra.Command {
        Use:   "summary",
        Short: "summarize the output of model(s)",
        Long:  summaryLongDescription,
-		Run:   summary,
+		RunE:   summary,
    }

    // Used for Summary
```

</details>


From bbr's standpoint, the main thing it needs from this change is to be able to detect and pull the errors (perhaps just for a subset of the models) from the JSON on stdout.  The above change results in the error message coming through on stdout alongside the JSON, which would cause bbr's to fail when reading in the JSON:

```
$ bbi nonmem summary --json integration/testdata/bbi_summary/66/66.lst integration/testdata/bbi_summary/66/66.ctl >out 2>err
$ tail out
            "success": false,
            "error_msg": "Must provide path to .lst (or .res) file for summary but provided 'integration/testdata/bbi_summary/66/66.ctl'\nCan also pass no extension and summary will infer .lst extension."
        }
    ],
    "Errors": [
        1,
        0
    ]
}
summarizing 2 model(s) failed
```

Just as a side remark, I was also surprised to see 1) the repeated error message to stderr (rather than it just being on stderr) and 2) the help message output on stderr:

```
$ head -5 err
Error: summarizing 2 model(s) failed
Usage:
  bbi nonmem summary [flags]

Flags:
```

So my (not very informed) assessment at this point is that switching `summary` to `RunE` sounds like a good idea conceptually but would need more work, presumably in upstream bbi code.

Reply:
That's easily fixed I believe by building the logger and config processing earlier. This way the execution layer that calls the cobra command already has the same logger and would just process logs the same way. Worth a discussion some time next week if you want

Reply:
@shairozan

> That's easily fixed I believe by building the logger and config processing earlier.  This way the execution layer that calls the cobra command already has the same logger and would just process logs the same way.

Hmm, I might be missing something, but for the duplicate print, the culprit on the bbi side is a plain `fmt.Println` call (see
[proposed fix in reclean pr](https://github.com/metrumresearchgroup/bbi/pull/276/commits/f5d4d9eb54b6036084c50968584f9bbe12751fd1)), so I don't think the logger would come into play.

And for the help output, it looks like `SilenceUsage` that I mentioned above would work.

> Worth a discussion some time next week if you want

Sure, I'd appreciate any pointers, especially if something I'm saying above is off the mark.  Thanks.